pkg/converters/types/interfaces.go

```diff
@@ -99,11 +103,12 @@ const (
 	ResourceGatewayClass ResourceType = "GatewayClass"
 	ResourceHTTPRoute    ResourceType = "HTTPRoute"
 
-	ResourceConfigMap ResourceType = "ConfigMap"
-	ResourceService   ResourceType = "Service"
-	ResourceEndpoints ResourceType = "Endpoints"
-	ResourceSecret    ResourceType = "Secret"
-	ResourcePod       ResourceType = "Pod"
+	ResourceConfigMap      ResourceType = "ConfigMap"
+	ResourceService        ResourceType = "Service"
+	ResourceEndpoints      ResourceType = "Endpoints"
+	ResourceEndpointSlices ResourceType = "EndpointSlices"
```

Comment:
I think you don't need this. Let me describe in a nutshell how tracking currently works:

- resource informer (`listers.go`) notifies the cache that a resource changed (create/update/delete)
- cache adds the resource identifier (ns+name) in an array of changed objects
- converter reads that array in order to find what changed

on the other side:

- object references (ing references svc, svc ref. endpoint, ing ref. secret, etc.) are tracked via `tracker.RefName()` (see below)
- `tracker.QueryLinks()` knows who should be synced again due to a resource update (see below)

`tracker.RefName()`:

https://github.com/jcmoraisjr/haproxy-ingress/blob/b87f94b9149bad3cef902e7e058c62ad2d83e02e/pkg/converters/ingress/ingress.go#L780-L783

`tracker.QueryLinks()`:

https://github.com/jcmoraisjr/haproxy-ingress/blob/b87f94b9149bad3cef902e7e058c62ad2d83e02e/pkg/converters/ingress/ingress.go#L219-L226

That said, what we need to do is to create a relation between svc and ep, so svc (hence, its ingress) will be synced when ep changes. The problem with endpointslices is that you cannot track the resource name because controller manager can create new endpointslices with unpredictable names.

My idea here is, instead, reuse all the former endpoint types - resource type and array of changed resources - and update it whenever an endpointslice is changed. Note that, instead of using its name, use it's annotation, in order to have a predictable name, btw the same name pattern the former ep already uses (it matches the svc name). You can test this by simply scaling up (provided you have free slots) or down, you should see that the current approach isn't changing backend servers, but the former ep does, and that proposed refactor should do it as well.

Reply:
Thanks for the explanation Joao. I've updated to reuse endpoint datastructures.